kallsyms/kallsyms.go

```diff
@@ -0,0 +1,446 @@
+	start libpf.Address
+	end   libpf.Address
+	mtime time.Time
+
+	buildID string
+	fileID  libpf.FileID
+	names   []byte
+	symbols []symbol
+}
+
+// Symbolizer provides the main API for reading, updating and querying
+// the kernel symbols.
+type Symbolizer struct {
+	valid atomic.Bool
+
+	modules []Module
+}
+
+// NewSymbolizer creates and returns a new kallsyms symbolizer.
+func NewSymbolizer() *Symbolizer {
+	return &Symbolizer{}
+}
+
+// addName appends the 'name' to the module's string slice, and returns
+// an index suitable for storing in the `symbol` struct.
+func (m *Module) addName(name string) uint32 {
+	index := len(m.names)
+	l := len(name)
+	m.names = append(m.names, byte(l))
+	m.names = append(m.names, stringutil.String2ByteSlice(name)...)
+	return uint32(index)
+}
+
+// bytesAt recovers a []byte representation of the string at `index`
+// received from previous `addName` call.
+func (m *Module) bytesAt(index uint32) []byte {
+	i := int(index)
+	return m.names[i+1 : i+1+int(m.names[i])]
+}
+
+// stringAt recovers the string at `index` received from previous `addName` call.
+func (m *Module) stringAt(index uint32) string {
+	return stringutil.ByteSlice2String(m.bytesAt(index))
+}
+
+// getSysfsMtime determine the kernel modules load time.
+func getSysfsMtime(mod string) time.Time {
+	if mod == Kernel {
+		return time.Time{}
+	}
+	if info, err := os.Stat(path.Join("/sys/module", mod)); err == nil {
+		return info.ModTime()
+	}
+	return time.Time{}
+}
+
+// parseSysfsUint reads a kernel module specific attribute from sysfs.
+func parseSysfsUint(mod, knob string) (uint64, error) {
+	text, err := os.ReadFile(path.Join("/sys/module", mod, knob))
+	if err != nil {
+		return 0, err
+	}
+	return strconv.ParseUint(strings.Trim(stringutil.ByteSlice2String(text), "\n"), 0, pointerBits)
+}
+
+// init prepare Module. The module name is recorded at index 0, and the needed
+// metadata (start, end, buildID) are read from sysfs.
+func (m *Module) init(mod string) {
+	// Record module name
+	m.addName(mod)
+
+	// Determine notes location and module size
+	notesFile := "/sys/kernel/notes"
+	if mod != Kernel {
+		notesFile = path.Join("/sys/module", mod, "notes/.note.gnu.build-id")
+
+		if addr, err := parseSysfsUint(mod, "sections/.text"); err == nil {
+			m.start = libpf.Address(addr)
+			if size, err := parseSysfsUint(mod, "coresize"); err == nil {
+				m.end = m.start + libpf.Address(size)
+			}
+		}
+	}
+
+	// Require at least 16 bytes of BuildID to ensure there is enough entropy for a FileID.
+	// 16 bytes could happen when --build-id=md5 is passed to `ld`. This would imply a custom
+	// kernel.
+	var err error
+	m.buildID, err = pfelf.GetBuildIDFromNotesFile(notesFile)
+	if err == nil && len(m.buildID) >= 16 {
+		m.fileID = libpf.FileIDFromKernelBuildID(m.buildID)
+	}
+}
+
+// finish will finalize the Module. The 'names' and 'symbols' slices are cloned
+// and sorted. A fallback fileID is synthesized if buildID is not available.
```

Comment:
Clarify why `Clone` is being used.

```suggestion
// to reduce memory use and sorted. A fallback fileID is synthesized if buildID
// is not available.
```

Reply:
Moved cloning back to the calling function because it needed to do some other buffer suffling. Better to keep those together. Added comments on all of them.